// react-marquease/umd/react-marquease.js
/*! react-marquease !*/
(function (global, factory) {
  typeof exports === 'object' && typeof module !== 'undefined' ? module.exports = factory(require('react')) :
  typeof define === 'function' && define.amd ? define(['react'], factory) :
  (global = typeof globalThis !== 'undefined' ? globalThis : global || self, global["react-marquease"] = factory(global.React));
})(this, (function (React) { 'use strict';

  function _extends() {
    _extends = Object.assign ? Object.assign.bind() : function (target) {
      for (var i = 1; i < arguments.length; i++) {
        var source = arguments[i];
        for (var key in source) {
          if (Object.prototype.hasOwnProperty.call(source, key)) {
            target[key] = source[key];
          }
        }
      }
      return target;
    };
    return _extends.apply(this, arguments);
  }

  function _arrayWithHoles(arr) {
    if (Array.isArray(arr)) return arr;
  }

  function _iterableToArrayLimit(arr, i) {
    var _i = null == arr ? null : "undefined" != typeof Symbol && arr[Symbol.iterator] || arr["@@iterator"];
    if (null != _i) {
      var _s,
        _e,
        _x,
        _r,
        _arr = [],
        _n = !0,
        _d = !1;
      try {
        if (_x = (_i = _i.call(arr)).next, 0 === i) {
          if (Object(_i) !== _i) return;
          _n = !1;
        } else for (; !(_n = (_s = _x.call(_i)).done) && (_arr.push(_s.value), _arr.length !== i); _n = !0);
      } catch (err) {
        _d = !0, _e = err;
      } finally {
        try {
          if (!_n && null != _i["return"] && (_r = _i["return"](), Object(_r) !== _r)) return;
        } finally {
          if (_d) throw _e;
        }
      }
      return _arr;
    }
  }

  function _arrayLikeToArray(arr, len) {
    if (len == null || len > arr.length) len = arr.length;
    for (var i = 0, arr2 = new Array(len); i < len; i++) arr2[i] = arr[i];
    return arr2;
  }

  function _unsupportedIterableToArray(o, minLen) {
    if (!o) return;
    if (typeof o === "string") return _arrayLikeToArray(o, minLen);
    var n = Object.prototype.toString.call(o).slice(8, -1);
    if (n === "Object" && o.constructor) n = o.constructor.name;
    if (n === "Map" || n === "Set") return Array.from(o);
    if (n === "Arguments" || /^(?:Ui|I)nt(?:8|16|32)(?:Clamped)?Array$/.test(n)) return _arrayLikeToArray(o, minLen);
  }

  function _nonIterableRest() {
    throw new TypeError("Invalid attempt to destructure non-iterable instance.\nIn order to be iterable, non-array objects must have a [Symbol.iterator]() method.");
  }

  function _slicedToArray(arr, i) {
    return _arrayWithHoles(arr) || _iterableToArrayLimit(arr, i) || _unsupportedIterableToArray(arr, i) || _nonIterableRest();
  }

  function _objectWithoutPropertiesLoose(source, excluded) {
    if (source == null) return {};
    var target = {};
    var sourceKeys = Object.keys(source);
    var key, i;
    for (i = 0; i < sourceKeys.length; i++) {
      key = sourceKeys[i];
      if (excluded.indexOf(key) >= 0) continue;
      target[key] = source[key];
    }
    return target;
  }

  function _objectWithoutProperties(source, excluded) {
    if (source == null) return {};
    var target = _objectWithoutPropertiesLoose(source, excluded);
    var key, i;
    if (Object.getOwnPropertySymbols) {
      var sourceSymbolKeys = Object.getOwnPropertySymbols(source);
      for (i = 0; i < sourceSymbolKeys.length; i++) {
        key = sourceSymbolKeys[i];
        if (excluded.indexOf(key) >= 0) continue;
        if (!Object.prototype.propertyIsEnumerable.call(source, key)) continue;
        target[key] = source[key];
      }
    }
    return target;
  }

  var _excluded = ["children", "defaultOffset", "reverse", "pause", "speed"];

  /**
   * react-marquease
   */
  var Marquee = /*#__PURE__*/React.forwardRef(function (_ref, ref) {
    var children = _ref.children,
      _ref$defaultOffset = _ref.defaultOffset,
      defaultOffset = _ref$defaultOffset === void 0 ? 0 : _ref$defaultOffset,
      _ref$reverse = _ref.reverse,
      reverse = _ref$reverse === void 0 ? false : _ref$reverse,
      _ref$pause = _ref.pause,
      pause = _ref$pause === void 0 ? false : _ref$pause,
      _ref$speed = _ref.speed,
      speed = _ref$speed === void 0 ? 1 : _ref$speed,
      rest = _objectWithoutProperties(_ref, _excluded);
    var _useState = React.useState(false),
      _useState2 = _slicedToArray(_useState, 2),
      mounted = _useState2[0],
      setMounted = _useState2[1];
    var _useState3 = React.useState(null),
      _useState4 = _slicedToArray(_useState3, 2),
      clientHeight = _useState4[0],
      setClientHeight = _useState4[1];
    var _useState5 = React.useState(null),
      _useState6 = _slicedToArray(_useState5, 2),
      clientWidth = _useState6[0],
      setClientWidth = _useState6[1];
    var _useState7 = React.useState(defaultOffset),
      _useState8 = _slicedToArray(_useState7, 2),
      offset = _useState8[0],
      setOffset = _useState8[1];
    var childrenRef = React.useRef(null);
    var intervalRef = React.useRef(null);
    React.useEffect(function () {
      setMounted(true);
    }, []);
    React.useEffect(function () {
      if (mounted) {
        var _childrenRef$current3, _childrenRef$current4;
        var handleResize = function handleResize() {
          var _childrenRef$current, _childrenRef$current2;
          setClientHeight(((_childrenRef$current = childrenRef.current) === null || _childrenRef$current === void 0 ? void 0 : _childrenRef$current.clientHeight) || 0);
          setClientWidth(((_childrenRef$current2 = childrenRef.current) === null || _childrenRef$current2 === void 0 ? void 0 : _childrenRef$current2.clientWidth) || 0);
        };
        window.addEventListener('resize', handleResize);
        setClientHeight(((_childrenRef$current3 = childrenRef.current) === null || _childrenRef$current3 === void 0 ? void 0 : _childrenRef$current3.clientHeight) || 0);
        setClientWidth(((_childrenRef$current4 = childrenRef.current) === null || _childrenRef$current4 === void 0 ? void 0 : _childrenRef$current4.clientWidth) || 0);
        return function () {
          window.removeEventListener('resize', handleResize);
        };
      }
    }, [mounted]);
    React.useEffect(function () {
      var tick = function tick() {
        var direction = reverse ? 1 : -1;
        setOffset(function (prevState) {
          var _childrenRef$current5;
          var nextOffset = prevState + direction * speed;
          if (Math.abs(nextOffset) >= (((_childrenRef$current5 = childrenRef.current) === null || _childrenRef$current5 === void 0 ? void 0 : _childrenRef$current5.clientWidth) || 0)) {
            nextOffset = 0;
          }
          return nextOffset;
        });
      };
      clearInterval(intervalRef.current);
      if (!pause) {
        intervalRef.current = window.setInterval(tick, 50);
      }
      return function () {
        clearInterval(intervalRef.current);
      };
    }, [childrenRef, intervalRef, pause, reverse, speed]);
    var childClones = React.useMemo(function () {
      if (typeof window === 'undefined') return null;
      var cloneCount = Math.max(2, clientWidth ? Math.ceil(window.innerWidth / (clientWidth / 2)) : 0);
      var direction = reverse ? 1 : -1;
      return Array(cloneCount).fill(0).map(function (item, index) {
        return /*#__PURE__*/React.createElement("div", {
          key: "marquee-clone-".concat(index),
          "data-marquee-child": index,
          style: {
            position: 'absolute',
            top: 0,
            right: reverse ? 0 : 'auto',
            transform: "translateX(".concat(clientWidth * index * direction * -1, "px)")
          }
        }, children);
      });
    }, [children, clientWidth, reverse]);
    if (!mounted) return null;
    return /*#__PURE__*/React.createElement("div", _extends({
      ref: ref
    }, rest, {
      style: {
        position: 'relative',
        height: clientHeight,
        overflow: 'hidden'
      }
    }), /*#__PURE__*/React.createElement("div", {
      ref: childrenRef,
      style: {
        position: 'absolute',
        top: '-9999px',
        left: '-9999px'
      }
    }, children), /*#__PURE__*/React.createElement("div", {
      "data-qa": "scroller",
      style: {
        transform: "translateX(".concat(offset, "px)")
      }
    }, childClones));
  });
  Marquee.displayName = 'Marquee';

  return Marquee;

}));
/* Copyright 2023 - Ryan Hefner - https://www.ryanhefner.com */
