// md-editor-rt/lib/es/MdPreview.mjs
import { j as a } from "./chunks/jsx-runtime.mjs";
import { useState as H, useEffect as b } from "react";
import { u as P, a as y, b as M, C as N } from "./chunks/Editor.mjs";
import { b as T, c as K } from "./chunks/index.mjs";
import { p as i, d as e } from "./chunks/config.mjs";
import "./chunks/index2.mjs";
import "@vavt/util";
import "medium-zoom";
import "copy-to-clipboard";
import "markdown-it";
import "markdown-it-image-figures";
import "markdown-it-task-lists";
import "lru-cache";
import "codemirror";
import "@codemirror/view";
import "@codemirror/language-data";
import "@codemirror/lang-markdown";
import "@codemirror/commands";
import "@codemirror/state";
import "@codemirror/language";
import "@lezer/highlight";
import "@codemirror/autocomplete";
const Z = (o) => {
  const {
    modelValue: d = e.modelValue,
    theme: n = e.theme,
    className: s = e.className,
    editorId: m = e.editorId,
    showCodeRowNumber: l = e.showCodeRowNumber,
    previewTheme: p = e.previewTheme,
    noMermaid: u = e.noMermaid,
    noKatex: h = e.noKatex,
    onHtmlChanged: g = e.onHtmlChanged,
    onGetCatalog: c = e.onGetCatalog,
    sanitize: f = e.sanitize,
    mdHeadingId: v = e.mdHeadingId,
    noIconfont: w = e.noIconfont,
    noHighlight: x = e.noHighlight
  } = o, [t] = H(() => ({
    editorId: m,
    noKatex: h,
    noMermaid: u,
    noIconfont: w,
    noHighlight: x
  }));
  P(t);
  const [C, I, r] = y(o);
  return b(() => () => {
    T.clear(m);
  }, []), /* @__PURE__ */ a(
    M.Provider,
    {
      value: {
        editorId: t.editorId,
        tabWidth: 2,
        theme: n,
        highlight: C,
        showCodeRowNumber: l,
        usedLanguageText: I,
        previewTheme: p
      },
      children: /* @__PURE__ */ a(
        "div",
        {
          id: t.editorId,
          className: K([
            i,
            s,
            n === "dark" && `${i}-dark`,
            r.fullscreen || r.pageFullscreen ? `${i}-fullscreen` : "",
            `${i}-previewOnly`
          ]),
          style: o.style,
          children: /* @__PURE__ */ a(
            N,
            {
              value: d,
              setting: r,
              mdHeadingId: v,
              onHtmlChanged: g,
              onGetCatalog: c,
              sanitize: f,
              noMermaid: t.noMermaid,
              noHighlight: t.noHighlight,
              noKatex: t.noKatex,
              formatCopiedText: o.formatCopiedText,
              previewOnly: !0
            },
            "preview-only"
          )
        }
      )
    }
  );
};
export {
  Z as default
};
