// md-editor-rt/lib/es/chunks/index.mjs
var L = Object.defineProperty;
var M = (o, t, e) => t in o ? L(o, t, { enumerable: !0, configurable: !0, writable: !0, value: e }) : o[t] = e;
var C = (o, t, e) => (M(o, typeof t != "symbol" ? t + "" : t, e), e);
import { a as S, j as b } from "./jsx-runtime.mjs";
import j, { useState as R, useMemo as A, useCallback as B, useEffect as P } from "react";
import { throttle as F } from "@vavt/util";
import { p as w, d as I } from "./config.mjs";
class q {
  constructor() {
    // 事件池
    C(this, "pools", {});
  }
  // 移除事件监听
  remove(t, e, n) {
    const s = this.pools[t] && this.pools[t][e];
    s && (this.pools[t][e] = s.filter((a) => a !== n));
  }
  // 清空全部事件，由于单一实例，多次注册会被共享内容
  clear(t) {
    this.pools[t] = {};
  }
  // 注册事件监听
  on(t, e) {
    return this.pools[t] || (this.pools[t] = {}), this.pools[t][e.name] || (this.pools[t][e.name] = []), this.pools[t][e.name].push(e.callback), this.pools[t][e.name].includes(e.callback);
  }
  // 触发事件
  emit(t, e, ...n) {
    this.pools[t] || (this.pools[t] = {});
    const s = this.pools[t][e];
    s && s.forEach((a) => {
      try {
        a(...n);
      } catch (r) {
        console.error(`${e} monitor event exception！`, r);
      }
    });
  }
}
const k = new q(), V = (o, t = "image.png") => {
  const e = o.split(","), n = e[0].match(/:(.*?);/);
  if (n) {
    const i = n[1], s = atob(e[1]);
    let a = s.length;
    const r = new Uint8Array(a);
    for (; a--; )
      r[a] = s.charCodeAt(a);
    return new File([r], t, { type: i });
  }
  return null;
}, W = (o) => {
  if (!o)
    return o;
  const t = o.split(`
`), e = ['<span rn-wrapper aria-hidden="true">'];
  return t.forEach(() => {
    e.push("<span></span>");
  }), e.push("</span>"), `<span class="code-block">${o}</span>${e.join("")}`;
}, _ = (o) => o.filter((t) => t !== !1).join(" "), G = (o, t) => {
  const e = o == null ? void 0 : o.getBoundingClientRect();
  if (t === document.documentElement)
    return e.top - t.clientTop;
  const n = t == null ? void 0 : t.getBoundingClientRect();
  return e.top - n.top;
}, X = () => `${Date.now().toString(36)}${Math.random().toString(36).substring(2)}`, H = ({
  tocItem: o,
  mdHeadingId: t,
  scrollElement: e,
  onClick: n,
  scrollElementOffsetTop: i = 0
}) => /* @__PURE__ */ S(
  "div",
  {
    className: _([
      `${w}-catalog-link`,
      o.active && `${w}-catalog-active`
    ]),
    onClick: (s) => {
      n && n(s, o), s.stopPropagation();
      const a = t(o.text, o.level, o.index), r = document.getElementById(a), d = e instanceof Element ? e : document.querySelector(e);
      if (r && d) {
        let c = r.offsetParent, p = r.offsetTop;
        if (d.contains(c))
          for (; c && d != c; )
            p += c == null ? void 0 : c.offsetTop, c = c == null ? void 0 : c.offsetParent;
        d == null || d.scrollTo({
          top: p - i,
          behavior: "smooth"
        });
      }
    },
    children: [
      /* @__PURE__ */ b("span", { title: o.text, children: o.text }),
      /* @__PURE__ */ b("div", { className: `${w}-catalog-wrapper`, children: o.children && o.children.map((s) => /* @__PURE__ */ b(
        H,
        {
          mdHeadingId: t,
          tocItem: s,
          scrollElement: e,
          onClick: n,
          scrollElementOffsetTop: i
        },
        s.text
      )) })
    ]
  }
), O = H, U = (o) => {
  const {
    editorId: t,
    mdHeadingId: e = I.mdHeadingId,
    theme: n = "light",
    offsetTop: i = 20
  } = o, [s, a] = R([]), [r, d] = R(), c = A(() => {
    const l = [];
    return s.forEach((E, m) => {
      const { text: T, level: h } = E, f = {
        level: h,
        text: T,
        index: m + 1,
        active: r === E
      };
      if (l.length === 0)
        l.push(f);
      else {
        let u = l[l.length - 1];
        if (f.level > u.level)
          for (let g = u.level + 1; g <= 6; g++) {
            const { children: v } = u;
            if (!v) {
              u.children = [f];
              break;
            }
            if (u = v[v.length - 1], f.level <= u.level) {
              v.push(f);
              break;
            }
          }
        else
          l.push(f);
      }
    }), l;
  }, [r, s]), [p] = R(() => o.scrollElement || `#${t}-preview-wrapper`), x = B(() => p instanceof HTMLElement ? p : document.querySelector(p), [p]);
  return P(() => {
    const l = F((h) => {
      if (h.length === 0)
        return a([]), !1;
      const { activeHead: f } = h.reduce(
        (u, g, v) => {
          const y = document.getElementById(
            e(g.text, g.level, v + 1)
          );
          if (y instanceof HTMLElement) {
            const N = x(), $ = G(y, N);
            if ($ < i && $ > u.minTop)
              return {
                activeHead: g,
                minTop: $
              };
          }
          return u;
        },
        {
          activeHead: h[0],
          minTop: Number.MIN_SAFE_INTEGER
        }
      );
      d(f), a(h);
    });
    k.on(t, {
      name: "catalogChanged",
      callback: l
    });
    const E = x(), m = E === document.documentElement ? window : E;
    k.emit(t, "pushCatalog");
    const T = () => {
      l(s);
    };
    return m == null || m.addEventListener("scroll", T), () => {
      k.remove(t, "catalogChanged", l), m == null || m.removeEventListener("scroll", T);
    };
  }, [i, s, e, x]), /* @__PURE__ */ b(
    "div",
    {
      className: `${w}-catalog${n === "dark" ? "-dark" : ""} ${o.className || ""} `,
      style: o.style,
      children: c.map((l) => /* @__PURE__ */ b(
        O,
        {
          mdHeadingId: e,
          tocItem: l,
          scrollElement: p,
          onClick: o.onClick,
          scrollElementOffsetTop: o.scrollElementOffsetTop
        },
        l.text
      ))
    }
  );
}, Y = j.memo(U);
export {
  Y as M,
  V as a,
  k as b,
  _ as c,
  W as g,
  X as u
};
